File: ios/App/IO/CsvLoader.swift
import Foundation

// CSV loader. Accepts comma- OR semicolon-separated (the two real-world
// dialects you hit when "Save as CSV" comes from US- vs DE-locale Excel).
// Handles double-quoted fields with escaped quotes.

public struct CsvPreview {
    public let headers: [String]
    public let rows: [[String]]
    public func column(_ idx: Int) -> [String] {
        rows.map { idx < $0.count ? $0[idx] : "" }
    }
}

public enum CsvLoader {
    public static func parse(_ data: Data) -> CsvPreview {
        let text = String(data: data, encoding: .utf8)
            ?? String(data: data, encoding: .isoLatin1)
            ?? ""
        let delim = detectDelimiter(text)
        let rows = tokenize(text, delim: delim)
        guard let first = rows.first, !rows.isEmpty else {
            return CsvPreview(headers: [], rows: [])
        }
        // Pad rows to the widest row.
        let cols = rows.map(\.count).max() ?? first.count
        let padded = rows.map { r -> [String] in
            r + Array(repeating: "", count: max(0, cols - r.count))
        }
        if looksLikeHeader(padded[0]) {
            let headers = padded[0].map { $0.isEmpty ? "(blank)" : $0 }
            return CsvPreview(headers: headers, rows: Array(padded.dropFirst()))
        } else {
            let headers = (1...cols).map { "col \($0)" }
            return CsvPreview(headers: headers, rows: padded)
        }
    }

    private static func detectDelimiter(_ text: String) -> Character {
        let head = text.prefix(8192)
        let commas = head.filter { $0 == "," }.count
        let semis  = head.filter { $0 == ";" }.count
        let tabs   = head.filter { $0 == "\t" }.count
        let scored: [(Character, Int)] = [(",", commas), (";", semis), ("\t", tabs)]
        return scored.max(by: { $0.1 < $1.1 })?.0 ?? ","
    }

    private static func looksLikeHeader(_ row: [String]) -> Bool {
        let nonEmpty = row.filter { !$0.isEmpty }
        guard !nonEmpty.isEmpty else { return false }
        let headerish = nonEmpty.filter { v in
            let lower = v.lowercased()
            let url = lower.hasPrefix("http://") || lower.hasPrefix("https://") || v.contains("://")
            let num = Double(v) != nil
            return !url && !num && v.count <= 64
        }.count
        return Double(headerish) / Double(nonEmpty.count) >= 0.6
    }

    // RFC-4180-ish tokenizer: handles double-quoted fields, escaped quotes,
    // and CR/LF line endings.
    private static func tokenize(_ text: String, delim: Character) -> [[String]] {
        var rows: [[String]] = []
        var current: [String] = []
        var field = ""
        var inQuotes = false
        var i = text.startIndex
        while i < text.endIndex {
            let c = text[i]
            if inQuotes {
                if c == "\"" {
                    let next = text.index(after: i)
                    if next < text.endIndex, text[next] == "\"" {
                        field.append("\"")
                        i = text.index(after: next); continue
                    }
                    inQuotes = false
                } else { field.append(c) }
            } else {
                switch c {
                case "\"":
                    inQuotes = true
                case delim:
                    current.append(field); field = ""
                case "\r":
                    // swallow; rely on \n to commit row
                    break
                case "\n":
                    current.append(field); field = ""
                    rows.append(current); current = []
                default:
                    field.append(c)
                }
            }
            i = text.index(after: i)
        }
        if !field.isEmpty || !current.isEmpty {
            current.append(field); rows.append(current)
        }
        return rows
    }
}

// MARK: - URL normalisation
public enum UrlNormalize {
    public static func normalize(_ raw: String) -> String? {
        let t = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                   .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
        guard !t.isEmpty else { return nil }
        let candidate: String
        let lower = t.lowercased()
        if lower.hasPrefix("http://") || lower.hasPrefix("https://") {
            candidate = t
        } else if t.hasPrefix("//") {
            candidate = "https:\(t)"
        } else if t.contains(" ") || t.contains("\n") {
            return nil
        } else if t.contains(".") && !t.hasPrefix(".") {
            candidate = "https://\(t)"
        } else { return nil }
        guard let url = URL(string: candidate), let s = url.scheme,
              (s == "http" || s == "https"), url.host != nil else { return nil }
        return url.absoluteString
    }
}

File: android/app/src/main/java/com/raban/etabli/probe/engine/CsvLoader.kt
// Copyright 2026 Raban Heller
// SPDX-License-Identifier: Apache-2.0

package com.raban.etabli.probe.engine

// CSV loader. Accepts comma OR semicolon (the two real-world "Save as CSV"
// dialects depending on locale). Handles double-quoted fields with escaped
// quotes. Returns the headers + rows so the UI can show a column picker.

data class CsvPreview(val headers: List<String>, val rows: List<List<String>>) {
    fun column(idx: Int): List<String> = rows.map { if (idx < it.size) it[idx] else "" }
}

object CsvLoader {
    fun parse(bytes: ByteArray): CsvPreview {
        val text = decodeText(bytes)
        val delim = detectDelimiter(text)
        val rows = tokenize(text, delim)
        if (rows.isEmpty()) return CsvPreview(emptyList(), emptyList())
        val cols = rows.maxOf { it.size }
        val padded = rows.map { r -> r + List(cols - r.size) { "" } }
        return if (looksLikeHeader(padded.first())) {
            val headers = padded.first().map { if (it.isBlank()) "(blank)" else it.trim() }
            CsvPreview(headers, padded.drop(1))
        } else {
            CsvPreview(List(cols) { "col ${it + 1}" }, padded)
        }
    }

    private fun decodeText(bytes: ByteArray): String =
        try { String(bytes, Charsets.UTF_8) } catch (_: Throwable) { String(bytes, Charsets.ISO_8859_1) }

    private fun detectDelimiter(text: String): Char {
        val head = text.take(8192)
        val commas = head.count { it == ',' }
        val semis  = head.count { it == ';' }
        val tabs   = head.count { it == '\t' }
        return listOf(',' to commas, ';' to semis, '\t' to tabs).maxByOrNull { it.second }?.first ?: ','
    }

    private fun looksLikeHeader(row: List<String>): Boolean {
        val nonEmpty = row.filter { it.isNotBlank() }
        if (nonEmpty.isEmpty()) return false
        val headerish = nonEmpty.count { v ->
            val lower = v.lowercase()
            val url = lower.startsWith("http://") || lower.startsWith("https://") || v.contains("://")
            val num = v.toDoubleOrNull() != null
            !url && !num && v.length <= 64
        }
        return headerish.toDouble() / nonEmpty.size >= 0.6
    }

    private fun tokenize(text: String, delim: Char): List<List<String>> {
        val rows = mutableListOf<List<String>>()
        var current = mutableListOf<String>()
        val field = StringBuilder()
        var inQuotes = false
        var i = 0
        while (i < text.length) {
            val c = text[i]
            if (inQuotes) {
                if (c == '"') {
                    if (i + 1 < text.length && text[i + 1] == '"') {
                        field.append('"'); i += 2; continue
                    }
                    inQuotes = false
                } else field.append(c)
            } else when (c) {
                '"' -> inQuotes = true
                delim -> { current += field.toString(); field.clear() }
                '\r' -> { /* swallow */ }
                '\n' -> { current += field.toString(); field.clear(); rows += current; current = mutableListOf() }
                else -> field.append(c)
            }
            i++
        }
        if (field.isNotEmpty() || current.isNotEmpty()) { current += field.toString(); rows += current }
        return rows
    }
}

object UrlNormalize {
    private val schemeRe = Regex("^https?://", RegexOption.IGNORE_CASE)
    fun normalize(raw: String): String? {
        val t = raw.trim().trim('"')
        if (t.isEmpty()) return null
        val candidate = when {
            schemeRe.containsMatchIn(t) -> t
            t.startsWith("//") -> "https:$t"
            t.contains(' ') || t.contains('\n') -> return null
            t.contains('.') && !t.startsWith('.') -> "https://$t"
            else -> return null
        }
        return try {
            val uri = java.net.URI(candidate)
            if (uri.scheme !in listOf("http", "https") || uri.host.isNullOrEmpty()) null
            else uri.toString()
        } catch (_: Throwable) { null }
    }
}
